# VEKTOR
### 21 feb 2022

matematika tapi kita akan banyak bahas di numpy

objectif
- basic understanding vector
- bisa mengoprasikan vektor
- implementasi vektor
--------------------------------------------------------------------------------

vektor merupakan besaran yang memiliki nilai dan arah

klo di data kita 
1 row adalah 1 vektor, atau setiap entry data adalah vektor

- kita bisa anggap vektor adalah satu nilai dari numbers
-------------------------------------------------------------------------------
kita akan coba mendefinisikan vektor menggunakan numpy array


In [1]:
import numpy as np

In [2]:
np.array([1,2,3])

array([1, 2, 3])

kalo kita pakai list dia tidak bisa melakukan oprasi aritmatika

In [4]:
list_1 = [1,2,3]
list_2 = [4,5,6]
list_1+list_2

[1, 2, 3, 4, 5, 6]

value nya tidak bertambah tapi yang terjadi adalah append (value nya di belakang)

------------------------------------------------------------------------------

sekarang kita coba sebagai array

In [5]:
arr_1= np.array(list_1)
arr_2= np.array(list_2)

arr_2+arr_1

array([5, 7, 9])

hasilnya berbeda jadi pada numpy yang kta butuhkan karena valuenya yang berubah

In [6]:
arr_1[-1] #kita akses yang paling belakang

3

In [7]:
arr_1[1] #ambil yang tengah

2

In [8]:
arr_1.size #melihat ada berapa element pada array

3

## vector norm

- pada konsep matematika di sebut dengan panjang vektor

np.linalg.norm(v)

In [10]:
print(arr_1,arr_2)

[1 2 3] [4 5 6]


In [11]:
#vectornorm dari arr1
np.linalg.norm(arr_1)

3.7416573867739413

In [12]:
#vectornorm dari arr2
np.linalg.norm(arr_2)

8.774964387392123

In [15]:
#default euclidean
arr = np.array([3,4])
np.linalg.norm(arr)
#konsepnya seperti pytagoras

5.0

bisa di sebut dengan 
L2 atau euclidean distance (pake pitagoras) 3,4 =5

-------------------------------------------------------------------------------------------------

L2 atau manhatan distance, ini di jumlahkan berdasarkan sumbunya (penjumblahan biasa) 3+4 =7


In [16]:
#discance manhatan
np.linalg.norm(arr,1)

7.0

In [17]:
#disrtance euclidean
np.linalg.norm(arr,2)

5.0

vektor norm ada 2 cara
1. loop for (waktu komputasinya lama) 
- Computational Thinking:norm = 0for i in range( len( vector ) ):norm=0norm+=vector[ I ]**2norm=sqrt( norm 
2. sec fungction (disarankan yang ini)
- Using Numpyimportnumpyasnp# define vector a = array([1, 2, 3])# calculate norm v_norm= np.linalg.norm(a) print(v_norm) Vector Norm on CodeOutput:   3.74165738677

## Oprasi aritmatika


### penjumblajhan antara vektor dan skalar atau vektor dengan vektor

In [18]:
print(arr_1,arr_2)

[1 2 3] [4 5 6]


In [19]:
#penjumblahan vektor dengan skalar atau bilangan
arr_1+10

array([11, 12, 13])

seluruh isi vektornya akan ter tambah dengan angka skalar nya

In [20]:
#penjumlahan vektor dengan vektor
arr_1+arr_2

array([5, 7, 9])

vektornya akan menambahkan sesuai urutan elementnya

untuk penambahanya harus menggunakan jumlah element yang sama

In [23]:
arr_3=np.array([1,2,3,4,5])

arr_1+arr_3

ValueError: operands could not be broadcast together with shapes (3,) (5,) 

jika jumlah element oprasinya berbeda maka tidak bisa di oprasikan
### element yang di oprasikan harus sama yaaaaaa

---------------------------------------------------------------------------------------------------------

## pengurangan

In [24]:
#pengurangan vektor dan skalar
print(arr_1,arr_2)

[1 2 3] [4 5 6]


In [25]:
arr_1-10

array([-9, -8, -7])

masing-masing akan di kurangin di seluruh elementnya

----------------------------------------------------------------------

sekarang kita coba antara vektor dan vektor

In [26]:
arr_1-arr_2

#yang terjadi adalah pengurangan sesuai poisis

array([-3, -3, -3])

## Pembagiann

In [27]:
#pembagian vektor dengan bilangan atau skalar

arr_1 / 2

#setiap element akan di bagi 2

array([0.5, 1. , 1.5])

In [29]:
print (arr_1,arr_2)

[1 2 3] [4 5 6]


In [28]:
#pembagian vektor dengan vektor akan ke element wise
arr_1/arr_2

array([0.25, 0.4 , 0.5 ])

# perkalian
- disini ad yg berbedaa

----------------------------------------------------------------------------
ada 3 cara
- mullti by scalar atau bilangan
- by vector
- dot prouct

In [30]:
print(arr_1,arr_2)

[1 2 3] [4 5 6]


###  scalar dengan vektor


In [31]:
arr_1 * 2

array([2, 4, 6])

In [33]:
arr_3 * 10
#setiap bilangan akan di kalikan dengan bilangan yg sama

array([10, 20, 30, 40, 50])

In [34]:
#kalo vektor dengan vektor dia element wise

arr_1*arr_2

array([ 4, 10, 18])

### dot product bberbeda karena hasilnya merupakan sekalar, dan jika di aplikasikan kita akan mendapatkan sudut setiap vektor agar bisa mendapatkan sudut apakah setiap vektor mirip sudutnya

simbol dot product adalah '@', setelah kali element wise setelah itu dia akan menjumblahkan dan hasilnya adalah bilangan biasa bukan berbentuk vektor lagi

terkait cara penulisan ad beberapa

In [35]:
#dot product
arr_1 @ arr_2


32

In [36]:
arr_1.dot(arr_2)

32

In [37]:
np.dot(arr_1,arr_2)

32

dengan memanfaatkan do tporduct ini bisa mendapatkan sudut tiap vektor yang bis merepretasikan kemiripan atau sudut vektor

||a|| adalah vektor norm a

a.b= ||a|| ||b|| cos teta nya atau sudut
lantak laaaaaaaaaa

yang biasa di manfaatkan dari rumus ini adalah cos tetanya maka bisa di modifikasi

cos teta = a @ b / ||a|| ||b||


semakin kecil sudutnya maka vektor ini bisa di bilang semakin miri p atau berdekatan

skalanya 0 - 1 banyak koma juga dong

- cosine similarity

In [39]:
print(arr_1,arr_2)

[1 2 3] [4 5 6]


In [42]:
cos_theta = arr_1 @ arr_2 / (np.linalg.norm(arr_1)* np.linalg.norm(arr_2))
cos_theta
#pastikan kembali tanda kurungny klo lupa beda hasilnya

0.9746318461970762

kita akan cob ajika vektornya itu ada di satu sumbu yang sama
- v1=3,0
- v2=5,0
- vektor= sumbu x , sumbu y


In [45]:
arr_3 = np.array([3,0])
arr_4 = np.array([5,0])

In [46]:
cos_theta = arr_3 @ arr_4 / (np.linalg.norm(arr_3)* np.linalg.norm(arr_4))
cos_theta
#pastikan kembali tanda kurungny klo lupa beda hasilnya

1.0

kita akan mencoba mencari sudut90derajat dimana cos = 0

In [49]:
arr_5 = np.array([0,4])
arr_6 = np.array([8,0])
cos_theta = arr_5 @ arr_6 / (np.linalg.norm(arr_5)* np.linalg.norm(arr_6))
cos_theta
#pastikan kembali tanda kurungny klo lupa beda hasilnya
# ingat dia itu adalah sumbu X , sumbu Y

0.0

sampai sini baru dapat COS teta nya sekarang kita akan mencari sudut teta

di numpy atau python tidak bisa langsung rubah ke sudut harus 2 langkah

harus ke radiant dulu baru ke degree

In [51]:
#cara mencari sudut teta
#merubah dalam bentuk radian
rad = np.arccos(cos_theta)
rad

1.5707963267948966

In [53]:
#convert radian ke degereee
degree = np.degrees(rad)
degree

90.0

sesi berakhir setelh ini kita akan cob acontoh kasus terkait penerapan vektor norm ini

----------------------------------------

## case study

memanfaatkan singularity apakah kalimat mirip atau tidak

- kalimat1 = saya pergi ke rumah nenek saya
- kalimmat2 = saya pergi ke rumah kakek saya
-             saya  pergi ke  rumah nenek kakek
-              2     1    1   1     1       1 saya lagi di awal uda di tulis
kal 1 kakek 0 
kal 2 nenek 0

caranya kita rubah ke bentuk vektor dengan cara maping dengan kata uniq yang ada

- saya
- pergi
- ke
- rumah
- nenek
- kakek

kal_1 = [2,1,1,1,1,0]
kal_2 = [2,1,1,1,0,1]


ini adalah list kalimat unik, 
selanjutnya kita kan petakan muncul berapa kali dan muncul aapa aja

misal di kalimat saya muncul 2x
pergi muncul 1x dst 
pokonya di kalimat pertama

setelah itu dia akan melihatdi kalimat ke 2 berapa kali muncul kalimat" nya

In [54]:
#caranya kita rubah ke bentuk vektor dengan cara maping dengan kata uniq yang ada

kal_1 = [2,1,1,1,1,0]
kal_2 = [2,1,1,1,0,1]

kal_1 = np.array(kal_1)

kal_2 = np.array(kal_2)

print(kal_1,kal_2)

[2 1 1 1 1 0] [2 1 1 1 0 1]


In [56]:
cos_theta = kal_1 @ kal_2 /(np.linalg.norm(kal_1)*np.linalg.norm(kal_2))
cos_theta

0.8749999999999998

bisa di bilang kalimat di atas ini adalah mirip
0.5 ke atas mirip
0.5 le bawah tidak miirip

case study 2
kalimat 1 = saya kemarin pergi ke medan
kalimat 2 = ibu saya sedang belanja

sekarang kita akan mapping dulu
           '`kal 1           kal2`
- saya      1               1
- kemarin   1               0   
- pergi     1               0
- ke        1               0
- medan     0               1
- ibu       0               1
- sedang    0               1
- belanja   0               1

In [57]:
vek_1 = [1,1,1,1,1,0,0,0]
vek_2 = [1,0,0,0,0,1,1,1]
vek_1= np.array(vek_1)
vek_2= np.array(vek_2)
print (vek_1,vek_2)

[1 1 1 1 1 0 0 0] [1 0 0 0 0 1 1 1]


In [58]:
cos_theta = vek_1 @ vek_2 /(np.linalg.norm(vek_1)*np.linalg.norm(vek_2))
cos_theta

0.22360679774997896

secara perhitungan kalimat yang ada tidak lah mirip begitu karena di bawah 0.5

In [ ]:
# special coba ke grrafik kartesian 